### PR TITLE
contrib/tlp: new package (1.6.1)

### PR DESCRIPTION
**Pull request by subnut**
- [x] Write dinit service files
- [x] Split `tlp-rdw` to a separate package

### CODE REVIEW

File: contrib/tlp/files/tlp [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# tlp service
+
+type            = scripted
+after           = pre-network.target
```
Comment (q66) [inline contrib/tlp/files/tlp:4]: what are these `after` supposed to mean? how are they related to power management? especially referencing networkmanager by name seems just wrong

the service is also missing a good hard dependency, which makes it unreliable

File: contrib/tlp/template.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+build_style = "makefile"
+make_cmd = "gmake"
+hostmakedepends = ["gmake"]
+pkgdesc = "Simple package"
+maintainer = "Subhaditya Nath <sn03.general@gmail.com>"
+license = "GPL-2.0-or-later"
+url = "https://linrunner.de/tlp"
+source = f"https://github.com/linrunner/TLP/archive/refs/tags/{pkgver}.tar.gz"
+sha256 = "f7d013691a92ffcf42ef1648565dbc24a33202046d3c8138dad1963a3169a0f5"
+
```
Comment (q66) [inline contrib/tlp/template.py:13]: spurious newline

File: contrib/tlp/template.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
+pkgname = "tlp"
+pkgver = "1.6.1"
+pkgrel = 0
+build_style = "makefile"
+make_cmd = "gmake"
+hostmakedepends = ["gmake"]
+pkgdesc = "Simple package"
```
Comment (q66) [inline contrib/tlp/template.py:7]: what's this pkgdesc?

File: contrib/tlp/template.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+url = "https://linrunner.de/tlp"
+source = f"https://github.com/linrunner/TLP/archive/refs/tags/{pkgver}.tar.gz"
+sha256 = "f7d013691a92ffcf42ef1648565dbc24a33202046d3c8138dad1963a3169a0f5"
+
+# no tests
+options = ["!check"]
+
+
+def init_configure(self):
+    self.make_install_args += [
```
Comment (q66) [inline contrib/tlp/template.py:19]: there is no reason to put this in init_configure

File: contrib/tlp/patches/01-tlp-func-base-in.patch [hunk 5]
```diff
@@ -0,0 +1,26 @@
+--- tlp-func-base.in
```
Comment (q66) [inline contrib/tlp/patches/01-tlp-func-base-in.patch:1]: support for `flock -w` is coming in chimerautils 14.0

File: contrib/tlp/files/tlp (hunk 1 above)
Comment (subnut) [reply contrib/tlp/files/tlp:5]: idk.. it was there in the systemd .service file distributed with tlp...

> missing a good hard dependency

added.

File: contrib/tlp/template.py (hunk 3 above)
Comment (subnut) [reply contrib/tlp/template.py]: oops

resolved.

File: contrib/tlp/patches/01-tlp-func-base-in.patch (hunk 5 above)
Comment (subnut) [reply contrib/tlp/patches/01-tlp-func-base-in.patch:1]: Oh! niceee

btw when?

File: contrib/tlp/files/tlp (hunk 1 above)
Comment (subnut) [reply contrib/tlp/files/tlp:5]: I guess it also configures the network interfaces (if available) to be as power-saving as possible...?

**Comment by subnut** [comment]
wait no it should clash with power-profiles-daemon

**Comment by subnut** [comment]
also i should probably wait till chimerautils 14 is released

File: contrib/hdparm/template.py [hunk 6]
```diff
@@ -0,0 +1,20 @@
+pkgname = "hdparm"
```
Comment (q66) [inline contrib/hdparm/template.py:1]: this does not belong in this PR does it

File: contrib/tlp/files/tlp [hunk 7]
```diff
@@ -0,0 +1,7 @@
+# tlp service
+
+type            = scripted
+depends-on      = local.target
+waits-for       = pre-network.target
```
Comment (q66) [inline contrib/tlp/files/tlp:5]: probably network.target

File: contrib/tlp/template.py [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+    "-j1",
+    "TLP_SBIN=/usr/bin",
+    "TLP_ULIB=/usr/lib/udev",
+    "TLP_ELOD=/usr/libexec/elogind/system-sleep",
+    "TLP_NO_INIT=1",
+    "TLP_WITH_ELOGIND=1",
+    "TLP_WITH_SYSTEMD=0",
+]
+hostmakedepends = ["gmake"]
+depends = ["perl", "ethtool", "chimerautils>=14"]
```
Comment (q66) [inline contrib/tlp/template.py:16]: the constrained dependency is pretty pointless

File: contrib/tlp/patches/00-Makefile.patch [hunk 9]
```diff
@@ -0,0 +1,123 @@
+--- Makefile
++++ Makefile
+@@ -141,44 +141,51 @@
+ 
+ install-tlp: all
```
Comment (q66) [inline contrib/tlp/patches/00-Makefile.patch:5]: you should probably submit these changes upstream because rewriting these patches is a pain

File: contrib/tlp/patches/01-fix-tlp-stat.patch [hunk 10]
```diff
@@ -0,0 +1,24 @@
+diff --git a/func.d/tlp-func-stat b/func.d/tlp-func-stat
+index 0eefb0f..a19bffd 100644
+--- a/func.d/tlp-func-stat
++++ b/func.d/tlp-func-stat
+@@ -14,7 +14,7 @@ readonly SESTATUS=sestatus
+ readonly SMARTCTL=smartctl
+ 
+ readonly RE_AC_QUIRK='^UNDEFINED$'
+-readonly RE_ATA_ERROR='ata[0-9]+: SError: {.*CommWake }'
++readonly RE_ATA_ERROR='ata[0-9]+: SError: \{.*CommWake \}'
```
Comment (q66) [inline contrib/tlp/patches/01-fix-tlp-stat.patch:10]: what's the purpose of this change?

File: contrib/hdparm/template.py (hunk 6 above)
Comment (subnut) [reply contrib/hdparm/template.py:1]: ah... forgot to add it to `depends`. Sorry!

File: contrib/tlp/patches/01-fix-tlp-stat.patch (hunk 10 above)
Comment (subnut) [reply contrib/tlp/patches/01-fix-tlp-stat.patch:10]: `RE_ATA_ERROR` is supposed to be an extended regexp.

According to [the POSIX spec](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap09.html#tag_09_05_02), unescaped `{` should be followed with a valid `DUP_COUNT`, otherwise results are undefined.

Reading the code, I've deduced that the intention of the braces in this regexp was to match the literal characters themselves. So, I fixed the regexp accordingly.

File: contrib/hdparm/template.py (hunk 6 above)
Comment (subnut) [reply contrib/hdparm/template.py:1]: done.

**Comment by Natthias** [comment]
Instead of packaging all of tlp, perhaps we should have a tlp-rdw subpackage. This is what most other distributions do and seems reasonable to me, as tlp-rdw is the only part of tlp which depends on NetworkManager, and is only a small subset of what tlp provides.

**Comment by q66** [comment]
ah, so that's why the systemd service is set to start after networkmanager...

perhaps we should put an `after = networkmanager` in there anyway then, i believe dinit should ignore it if unavailable or not set to start

File: contrib/hdparm/patches/00-fix-Makefile.patch [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+ static:
+ 	LDFLAGS=-static $(MAKE)
+@@ -52,9 +51,9 @@
+ 	if [ -f $(DESTDIR)$(sbindir)/hdparm ]; then rm -f $(DESTDIR)$(sbindir)/hdparm ; fi
+ 	if [ -f $(DESTDIR)$(mandir)/man8/hdparm.8 ]; then rm -f $(DESTDIR)$(mandir)/man8/hdparm.8 ;\
+ 	elif [ -f $(DESTDIR)$(oldmandir)/man8/hdparm.8 ]; then rm -f $(DESTDIR)$(oldmandir)/man8/hdparm.8 ; fi
+-	$(INSTALL_PROGRAM) -D hdparm $(DESTDIR)$(sbindir)/hdparm
+-	if [ -d $(DESTDIR)$(mandir) ]; then $(INSTALL_DATA) -D hdparm.8 $(DESTDIR)$(mandir)/man8/hdparm.8 ;\
+-	elif [ -d $(DESTDIR)$(oldmandir) ]; then $(INSTALL_DATA) -D hdparm.8 $(DESTDIR)$(oldmandir)/man8/hdparm.8 ; fi
++	$(INSTALL_PROGRAM) hdparm $(DESTDIR)$(sbindir)/hdparm
```
Comment (q66) [inline contrib/hdparm/patches/00-fix-Makefile.patch:29]: i believe this should not need patching with sbindir set in template?

File: contrib/hdparm/template.py [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+hostmakedepends = ["gmake"]
+makedepends = ["linux-headers"]
+pkgdesc = "Get/Set ATA/SATA drive parameters under Linux"
+maintainer = "Subhaditya Nath <sn03.general@gmail.com>"
+license = "BSD-1-Clause AND GPL-2.0-only"
+url = "https://sourceforge.net/projects/hdparm"
+source = f"$(SOURCEFORGE_SITE)/{pkgname}/{pkgname}-{pkgver}.tar.gz"
+sha256 = "d14929f910d060932e717e9382425d47c2e7144235a53713d55a94f7de535a4b"
+# no test suite
+options = ["!check"]
```
Comment (q66) [inline contrib/hdparm/template.py:16]: can you try building it with CFI on? and test it a bit

File: contrib/hdparm/patches/00-fix-Makefile.patch (hunk 11 above)
Comment (q66) [reply contrib/hdparm/patches/00-fix-Makefile.patch:29]: ah, nevermind, the -D still needs patching

**Comment by flukeeey** [comment]
Bump.

Would love to be able to use TLP on Chimera too.

**Comment by subnut** [comment]
Sorry, kinda caught up in real-life situations... can't pick this up right now..

**Comment by flukeeey** [comment]
No worries, I can try to pick this up.

**Comment by subnut** [comment]
Hey @q66 i see that you made `power-profiles-daemon` conflict with `tlp`

Should I add something similar to the template of this package too?

**Comment by subnut** [comment]
Also, I've tried to split all the tlp-rdw stuff into a separate tlp-rdw package.

If I've missed something, lemme know @Natthias

File: contrib/hdparm/template.py (hunk 12 above)
Comment (subnut) [reply contrib/hdparm/template.py:16]: Is that necessary? These are all text files....

Comment (subnut) [reply contrib/hdparm/template.py:16]: I don't think anything's being compiled...? So how'd CFI help?

sorry.... i am a total noob when it comes to hardening stuff like CFI, LTO, sanitization and whatever else is there...

**Comment by subnut** [comment]
Sorry for taking so long... endsem exams were going on, and some other stuff was pretty messed up..

File: contrib/hdparm/template.py (hunk 12 above)
Comment (q66) [reply contrib/hdparm/template.py:16]: what do you mean? hdparm is a C program

File: contrib/tlp/files/tlp [hunk 13]
```diff
@@ -0,0 +1,7 @@
+# tlp service
+
+type            = scripted
+depends-on      = local.target
+waits-for       = network.target
```
Comment (q66) [inline contrib/tlp/files/tlp:5]: as per the discussion in this PR, this should have `after = networkmanager` in it

Comment (subnut) [reply contrib/tlp/files/tlp:5]: .... no?

@Natthias
> ... tlp-rdw is the only part of tlp which depends on NetworkManager, and is only a small subset of what tlp provides.

and tlp-rdw has been split to a separate package.

---

I did comb through the code just to look if `tlp init start` invokes anything from `tlp-rdw` ... and it doesn't seem like it does...


IIRC it seemed to me that all the tlp-rdw stuff was invoked by the  
`usr/lib/NetworkManager/dispatcher.d/99tlp-rdw-nm`

...idk :shrug:

Comment (subnut) [reply contrib/tlp/files/tlp:5]: @Natthias ?

Comment (Natthias) [reply contrib/tlp/files/tlp:5]: Yea, tlp-rdw doesn't seem to actually need to have tlp running, so leaving out the `after = networkmanager` seems fine. I also tried testing it briefly and tlp-rdw seemed to work fine without tlp actually running, all it needed was networkmanager.

Bit odd that upstream has it start after networkmanager, so maybe there is a reason, but not one I can see either.

File: contrib/hdparm/template.py (hunk 12 above)
Comment (subnut) [reply contrib/hdparm/template.py:16]: Ah.. sorry... had a brainfart there 😅

Comment (subnut) [reply contrib/hdparm/template.py:16]: It does seem to be building with CFI enabled.... but I don't have any HDDs to test it on...

**Comment by subnut** [comment]
> Hey @q66 i see that you made `power-profiles-daemon` conflict with `tlp`
> 
> Should I add something similar to the template of this package too? Or is it enough to have a conflict in any one of the templates?
> 
> e.g.
> 
> ```python
> depends += ["!power-profiles-daemon"]
> ```

@q66 Should I?

File: contrib/tlp/template.py [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+    "TLP_SBIN=/usr/bin",
+    "TLP_ULIB=/usr/lib/udev",
+    "TLP_ELOD=/usr/libexec/elogind/system-sleep",
+    "TLP_NO_INIT=1",
+    "TLP_WITH_ELOGIND=1",
+    "TLP_WITH_SYSTEMD=0",
+]
+hostmakedepends = ["gmake"]
+depends = ["perl", "ethtool", "hdparm"]
+pkgdesc = "Optimize Linux Laptop Battery Life"
```
Comment (q66) [inline contrib/tlp/template.py:17]: Please Don't Randomly Title Case Things

File: contrib/tlp/template.py [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+options = ["!check"]
+
+
+def post_install(self):
+    self.install_service(self.files_path / "tlp")
+
+
+@subpackage("tlp-rdw")
+def _rdw(self):
+    self.depends = [f"tlp={pkgver}-r{pkgrel}", "networkmanager"]
```
Comment (q66) [inline contrib/tlp/template.py:33]: for orig-depends/install_if, `{pkgname}` should be used in place of the actual package name

also the subsequent install_if should not reference the depends by reference; either copy the list literal (better) or use `list(x)` (worse)

File: contrib/hdparm/template.py [hunk 16]
```diff
@@ -0,0 +1,21 @@
+pkgname = "hdparm"
+pkgver = "9.65"
+pkgrel = 0
+build_style = "makefile"
+make_cmd = "gmake"
+make_install_args = ["-j1", "binprefix=/usr", "sbindir=/usr/bin"]
+hostmakedepends = ["gmake"]
+makedepends = ["linux-headers"]
+pkgdesc = "Get/Set ATA/SATA drive parameters under Linux"
```
Comment (q66) [inline contrib/hdparm/template.py:9]: `Command-line interface to hard drive parameters`

File: contrib/tlp/template.py (hunk 14 above)
Comment (subnut) [reply contrib/tlp/template.py:17]: Oh I'd just copied that verbatim from https://linrunner.de/tlp/ 😂

Will fix

Comment (subnut) [reply contrib/tlp/template.py:17]: hmm... how about this?

```python
pkgdesc = "Battery life optimization and power management utility"
```

Comment (q66) [reply contrib/tlp/template.py:17]: something like `Battery life optimization utility` is enough

File: contrib/hdparm/template.py [hunk 17]
```diff
@@ -0,0 +1,21 @@
+pkgname = "hdparm"
+pkgver = "9.65"
+pkgrel = 0
+build_style = "makefile"
+make_cmd = "gmake"
+make_install_args = ["-j1", "binprefix=/usr", "sbindir=/usr/bin"]
+hostmakedepends = ["gmake"]
+makedepends = ["linux-headers"]
+pkgdesc = "Command-line interface to hard drive parameters"
```
Comment (q66) [inline contrib/hdparm/template.py:9]: hard disk drive

File: contrib/tlp/template.py [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+
+
+def post_install(self):
+    self.install_service(self.files_path / "tlp")
+
+
+@subpackage("tlp-rdw")
+def _rdw(self):
+    self.depends = [f"{pkgname}={pkgver}-r{pkgrel}", "networkmanager"]
+    self.install_if =  [f"{pkgname}={pkgver}-r{pkgrel}", "networkmanager"]
```
Comment (q66) [inline contrib/tlp/template.py:34]: extra space

File: contrib/hdparm/template.py (hunk 17 above)
Comment (subnut) [reply contrib/hdparm/template.py:9]: i'm stupid.

File: contrib/tlp/template.py (hunk 18 above)
Comment (subnut) [reply contrib/tlp/template.py]: hmm... the black command didn't run automatically for some reason...


time to debug that!

Comment (q66) [reply contrib/tlp/template.py]: i have no idea what "run automatically" is supposed to mean here
